Use Lazy deserialization for queries with nested compact & portable API-1553

srknzl opened the pull request:
`DeserializedGenericRecord`s were being used during queries of nested compact and portable fields. This behaviour for compact can be observed by debugging the test `CompactNestedFieldsTest.test_basicQuerying` and putting a breakpoint [here](https://github.com/hazelcast/hazelcast/blob/99b6c7463ee4b7cfc6313c31645ef09fbe27992b/hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/FieldOperations.java#L437). `genericRecord` variable will be a `DeserializedGenericRecord`.

I realized it comes from `readAsLeafObjectOnQuery` for the `FieldKind.COMPACT`. 
 
ColumnExpression will return DeserializedGenericRecord if a compact field is accessed because it is the actual column access operation. FieldAccessOperation will always return the lazy internal generic record. 

If a field is accessed like a.b.c.d 3 FieldAccessOperations will happen for d, c, and b; and 1 ColumnExpression will happen for a as the last operation.

Nested querying does not support Array/List or Map access operators in any form. And there is no array type. So we don't need to make lazy deserialization work for that use case.

Fixes https://hazelcast.atlassian.net/browse/API-1553

Breaking changes (list specific methods/types/messages):
* None?

Checklist:
- [x] Labels (`Team:`, `Type:`, `Source:`, `Module:`) and Milestone set
- [x] Label `Add to Release Notes` or `Not Release Notes content` set
- [x] Request reviewers if possible
- [x] Send backports/forwardports if fix needs to be applied to past/future releases
- [x] New public APIs have `@Nonnull/@Nullable` annotations
- [x] New public APIs have `@since` tags in Javadoc

hz-devops-test commented:
The job `Hazelcast-pr-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/3271/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/3271/Hazelcast-pr-compiler-3271.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</pre>
</details>

hz-devops-test commented:
The job `Hazelcast-pr-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/3394/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-compiler/3394/Hazelcast-pr-compiler-3394.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.2.0:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.38 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-compiler/hazelcast/src/main/java/com/hazelcast/query/impl/getters/Extractors.java:78: Line is longer than 130 characters (found 134). [LineLength]
--------------------------
</pre>
</details>

mdumandag commented:
@srknzl This is more like an SQL-related change, rather than serialization. We should definitely get a review from them

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/GenericRecordQueryReader.java:
```diff
@@ -250,6 +258,10 @@ private Object readLeaf(InternalGenericRecord record, String path) {
             return null;
         }
         FieldKind kind = record.getFieldKind(path);
+        if (kind == FieldKind.COMPACT && useLazyDeserialization) {
```
What about portable?

srknzl replied:
If we agree on the compact case, we can easily add portable as well. Will I need to add new tests for portable?  Is there a test like CompactStreamSerializerValueReaderQuickTest for portable that will test we don't break anything?

mdumandag replied:
I think that test was copied from the Portable counter-part, see `DefaultPortableReaderQuickTest`

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/GenericRecordQueryReader.java:
```diff
@@ -58,8 +58,16 @@ public final class GenericRecordQueryReader implements ValueReader {
+    private final boolean useLazyDeserialization;
+
     public GenericRecordQueryReader(InternalGenericRecord rootRecord) {
         this.rootRecord = rootRecord;
+        this.useLazyDeserialization = false;
```
minor
```suggestion
        this(rootRecord, false);
```

srknzl replied:
It is forgotten, fixed.

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/getters/CompactGetter.java:
```diff
@@ -30,18 +30,33 @@ public CompactGetter(InternalSerializationService serializationService) {
+        }
+        int depth = (int) metadata;
+        // Use lazy deserialization if we use a field of the nested field or a deeper access.
+        boolean useLazyDeserialization = depth > 1;
+        return getValueInternal(obj, attributePath, useLazyDeserialization);
```
But what if I do `SELECT (foo).bar.baz FROM xxx`, and `baz` is also a compact type? Now, if I see it correctly, you will return an internal generic record, but it should have been a deserialized generic record

srknzl replied:
Added a test about it to CompactNestedFieldsTest. It returns a deserialized generic record. I removed the depth logic as well. I think we don't need it. ColumnExpression will return DeserializedGenericRecord if a compact field is accessed because it is the actual column access operation. FieldAccessOperation will always return the lazy internal generic record.

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/13971/Hazelcast-pr-builder-13971.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/13971/Hazelcast-pr-builder-13971.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Errors: 
[ERROR]   PortableNestedFieldsTest.test_basicQuerying:114 ? HazelcastSql Unable to injec...
[INFO] 
[ERROR] Tests run: 5124, Failures: 0, Errors: 1, Skipped: 9
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

srknzl commented:
@viliam-durina @ivanthescientist Can you check this PR?

mdumandag commented on hazelcast/src/main/java/com/hazelcast/query/impl/getters/Extractors.java:
```diff
@@ -75,11 +75,16 @@ public Object extract(Object target, String attributeName, Object metadata) {
 
-    public Object extract(Object target, String attributeName, Object metadata, boolean failOnMissingReflectiveAttribute) {
+    public Object extract(Object target, String attributeName, Object metadata, boolean failOnMissingReflectiveAttribute,
+                          boolean useLazyDeserialization) {
```
can't we pass `useLazyDeserialization` as metadata here, instead of introducing a new parameter? I feel like metadata is put here for these kinds of situations

srknzl replied:
Done.

mdumandag reviewed (commented):
@srknzl can we also write a test that verifies we are indeed using lazy deserialization for field accesses?

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/14410/Hazelcast-pr-builder-14410.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/14410/Hazelcast-pr-builder-14410.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</pre>
</details>

srknzl commented on hazelcast/src/main/java/com/hazelcast/query/impl/getters/PortableGetter.java:
```diff
@@ -30,8 +30,7 @@ public PortableGetter(InternalSerializationService serializationService) {
 
-    @Override
-    public Object getValue(Object target, String fieldPath) throws Exception {
+    public Object getValueInternal(Object target, String fieldPath, boolean useLazyDeserialization) throws Exception {
```
Done.

srknzl commented on hazelcast/src/main/java/com/hazelcast/sql/impl/extract/GenericFieldExtractor.java:
```diff
@@ -42,11 +42,10 @@ public GenericFieldExtractor(
 
-    @Override
-    public Object get() {
+    public Object getInternal(boolean useLazyDeserialization) {
```
Done

srknzl commented:
Dear @ivanthescientist @mdumandag, while writing the SqlLazyDeserializationTest, I realized the lazy deserialization was not working as expected. Can you recheck the PR, because it is changed a lot. ?

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/14430/Hazelcast-pr-builder-14430.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/14430/Hazelcast-pr-builder-14430.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>

</pre>
</details>

srknzl commented:
**Summary**: 

The lazy deserialization support is added for only the following cases:

1. For ColumnExpression, when eval() is used, meaning the expression is not the top expression; and ColumnExpression.eval() is called by FieldAccessExpression. This means we will do a field access. If the field Column we will be accessing via ColumnExpression is a compact or a portable field, it will be returned as an InternalGenericRecord, which is not deserialized.
2. For FieldAccessExpression, only the latest field access will use eager deserialization for compact and portable. All the field access prior to the latest field access will be lazy.

This should ensure we use lazy deserialization for nested queries.

**API Changes:**

1. To Expression interface, I added a new **eval** method:

`T eval(Row row, ExpressionEvalContext context, boolean useLazyDeserialization)`

This will be called when an expression should be evaluated with lazy deserialization enabled. The default implementation of this method passes useLazyDeserialization as false, and reuse the current `eval(Row row, ExpressionEvalContext context,)`, meaning that lazy deserialization is disabled by default.

2. Added `get(useLazyDeserialization)` versions of `get()` method that exists for extractors. This boolean value is passed as metadata to CompactGetter or PortableGetter. Then the getter class use it to return a lazy compact/portable object.

hz-devops-test commented:
The job `Hazelcast-pr-EE-compiler` of your PR failed. (Hazelcast internal details: [build log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-EE-compiler/3484/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-EE-compiler/3484/Hazelcast-pr-EE-compiler-3484.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:animal-sniffer-maven-plugin:1.21:check (source-java8-check) on project hazelcast-jet-cdc-debezium: Failed to check signatures:  failed to process jar /home/jenkins/.m2/repository/org/lz4/lz4-java/1.8.0/lz4-java-1.8.0.jar : zip file is empty -> [Help 1]
--------------------------
</pre>
</details>

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java:
```diff
@@ -145,7 +145,8 @@ public class SerializationServiceV1 extends AbstractSerializationService {
     private final PortableContextImpl portableContext;
-    private final PortableSerializer portableSerializer;
+    // Not final for testing
+    private PortableSerializer portableSerializer;
```
And, of course, you have to use something like this to use this serialization service instead: https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/test/java/com/hazelcast/internal/serialization/impl/compact/schema/CompactSchemaReplicationTestBase.java#L94

srknzl replied:
We want to mock the portable serializer for sure because only one call to portableSerializer.readAsInternalGenericRecord comes from  `SerializationServiceV1.readAsInternalGenericRecord`. The others come from `PortableInternalGenericRecord.java` as `readAsInternalGenericRecord(BufferObjectDataInput in, int factoryId, int classId)`.

Regarding compact serializer, there is no problem because it was not final before.

mdumandag replied:
Then, what about returning a spied internal generic record here, and keeping track of readAsInternalGenericRecord calls over it?

srknzl commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/SqlLazyDeserializationTest.java:
```diff
@@ -0,0 +1,135 @@
+@RunWith(HazelcastSerialClassRunner.class)
+public class SqlLazyDeserializationTest {
+
+    private static final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
```
fixed

srknzl commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/SqlLazyDeserializationTest.java:
```diff
@@ -0,0 +1,135 @@
+
+    private static final TestHazelcastFactory hazelcastFactory = new TestHazelcastFactory();
+
+    HazelcastInstance instance;
+    HazelcastInstance client;
+
+    CompactStreamSerializer spyCompactStreamSerializer;
+    PortableSerializer spyPortableSerializer;
```
done.

srknzl commented on hazelcast-sql/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
I ran all the sql tests with MicrosoftJDK17, they passed fine. Only `com.hazelcast.it.CheckNoticeFileIT#verifyNoticeFile ` failed. I think it is unrelated failure and about my IDE configuration.

srknzl replied:
What about making PortableSerializer class nonfinal? Is it a must?

mdumandag replied:
I think we can do this, yes

srknzl commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/SqlLazyDeserializationTest.java:
```diff
@@ -0,0 +1,135 @@
+ * This test does not extend SqlTestSupport on purpose because we want full control over the member
+ * we start.
+ */
+@RunWith(HazelcastSerialClassRunner.class)
```
I think it is because they extend `SqlTestSupport`.

hz-devops-test commented:
The job `Hazelcast-pr-builder` of your PR failed. (Hazelcast internal details: [build log](
https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/14673/Hazelcast-pr-builder-14673.zip), [artifacts](https://s3.console.aws.amazon.com/s3/buckets/j-artifacts/Hazelcast-pr-builder/14673/Hazelcast-pr-builder-14673.zip)).
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   JobLifecycleMetricsTest.multipleJobsSubmittedAndCompleted:102->HazelcastTestSupport.assertTrueEventually:1349->HazelcastTestSupport.assertTrueEventually:1247->lambda$multipleJobsSubmittedAndCompleted$1:102->assertJobStats:187->assertJobStatsOnMember:196 expected:<2> but was:<1>
[INFO] 
[ERROR] Tests run: 50958, Failures: 1, Errors: 0, Skipped: 242
[INFO] 

[ERROR] There are test failures.
</pre>
</details>

srknzl commented:
run-lab-run

mdumandag commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/SqlLazyDeserializationTest.java:
```diff
@@ -0,0 +1,151 @@
+                .addLongField("id")
+                .addStringField("name")
+                .build();
+        serializationService.getPortableContext().registerClassDefinition(officeType);
```
why don't we configure these class definitions before the instance starts, through the serialization config?

mdumandag commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/SqlLazyDeserializationTestUtil.java:
```diff
@@ -0,0 +1,117 @@
+public class SqlLazyDeserializationTestUtil {
+    public static class SqlLazyDeserializationTestInstanceFactory extends TestHazelcastFactory {
+        public HazelcastInstance newHazelcastInstance(Config config) {
+
```
minor: unneeded newline

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/InternalSerializationService.java:
```diff
@@ -178,4 +180,13 @@ default <T> T readObject(ObjectDataInput in) {
 
     byte getVersion();
 
+    /**
+     * Used by tests
+     */
+    CompactStreamSerializer getCompactStreamSerializer();
+
+    /**
+     * Used by tests
+     */
+    PortableSerializer getPortableSerializer();
```
I assume these are no longer needed

srknzl replied:
getPortableSerializer is used in another test. So it is required. Removed the other one.

mdumandag replied:
But, we don't need to define that in this interface, right? Whatever that test is, I assume it can continue using the method defined in the implementations

srknzl replied:
I think I can remove it as well. Let me see.

mdumandag commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/PortableNestedFieldsTest.java:
```diff
@@ -59,37 +67,69 @@ public static void beforeClass() {
                 .build();
         serializationService.getPortableContext().registerClassDefinition(organizationType);
 
+        organizationAndLongType = new ClassDefinitionBuilder(1, 4)
```
why don't we configure these class definitions before the instance starts, through the serialization config?

mdumandag commented on hazelcast-sql/src/test/java/com/hazelcast/jet/sql/impl/type/SqlLazyDeserializationTestUtil.java:
```diff
@@ -0,0 +1,117 @@
+import static org.mockito.Mockito.spy;
+
+public class SqlLazyDeserializationTestUtil {
+    public static class SqlLazyDeserializationTestInstanceFactory extends TestHazelcastFactory {
```
All of this is used in a single test, right? And, it is pretty specific to that test. What do you think about moving all of these definitions to that test?

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/AbstractSerializationService.java:
```diff
@@ -719,6 +719,11 @@ protected void verifyDefaultSerializersNotOverriddenWithCompact() {
     }
 
+    // Used by tests
+    public CompactStreamSerializer getCompactStreamSerializer() {
```
I assume this is no longer needed

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/SerializationServiceV1.java:
```diff
@@ -365,7 +365,7 @@ private void registerClassDefinition(ClassDefinition cd, Map<Integer, Map<Intege
     }
 
-    public final PortableSerializer getPortableSerializer() {
+    public PortableSerializer getPortableSerializer() {
```
Let's bring this back

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableSerializer.java:
```diff
@@ -36,7 +36,7 @@
 import java.util.Set;
 
-public final class PortableSerializer implements StreamSerializer<Object> {
+public class PortableSerializer implements StreamSerializer<Object> {
```
I think we no longer need to do that

mdumandag commented on hazelcast/src/main/java/com/hazelcast/internal/serialization/impl/portable/PortableSerializer.java:
```diff
@@ -314,7 +314,8 @@ <T> T read(BufferObjectDataInput in, int factoryId, int classId) throws IOExcept
     }
 
-    <T> T readAsInternalGenericRecord(BufferObjectDataInput in, int factoryId, int classId) throws IOException {
+    // Public for testing
+    public <T> T readAsInternalGenericRecord(BufferObjectDataInput in, int factoryId, int classId) throws IOException {
```
I think now we can bring this to the old state

mdumandag commented on hazelcast/src/main/java/com/hazelcast/jet/impl/serialization/DelegatingSerializationService.java:
```diff
@@ -127,6 +128,11 @@ public PortableContext getPortableContext() {
         return delegate.getPortableContext();
     }
 
+    @Override
+    public PortableSerializer getPortableSerializer() {
+        return delegate.getPortableSerializer();
+    }
+
```
not needed anymore

mdumandag commented on hazelcast/src/test/java/com/hazelcast/test/compatibility/SamplingSerializationService.java:
```diff
@@ -244,6 +246,16 @@ public byte getVersion() {
         return delegate.getVersion();
     }
 
+    @Override
+    public CompactStreamSerializer getCompactStreamSerializer() {
+        return delegate.getCompactStreamSerializer();
+    }
+
+    @Override
+    public PortableSerializer getPortableSerializer() {
+        return delegate.getPortableSerializer();
+    }
+
```
not needed anymore
